sample-code/spring-app/src/main/java/com/sap/ai/sdk/app/services/OpenAiServiceV2.java

```diff
@@ -104,62 +93,31 @@ public OpenAiChatCompletionResponse chatCompletionImage(@Nonnull final String li
-    final var function =
-        new FunctionObject()
-            .name("weather")
-            .description("Get the weather for the given location")
-            .parameters(schemaMap);
-    final var tool = new ChatCompletionTool().type(FUNCTION).function(function);
+    final OpenAiClient client = OpenAiClient.forModel(GPT_4O_MINI);
 
     final var messages = new ArrayList<OpenAiMessage>();
     messages.add(OpenAiMessage.user("What's the weather in %s in %s?".formatted(location, unit)));
 
-    // Assistant will call the function
-    final var request = new OpenAiChatCompletionRequest(messages).withTools(List.of(tool));
-    final OpenAiChatCompletionResponse response =
-        OpenAiClient.forModel(GPT_4O_MINI).chatCompletion(request);
-
-    // 2. Optionally, execute the function.
-    final OpenAiAssistantMessage assistantMessage = response.getMessage();
-    messages.add(assistantMessage);
-
-    final OpenAiToolCall toolCall = assistantMessage.toolCalls().get(0);
-    if (!(toolCall instanceof OpenAiFunctionCall functionCall)) {
-      throw new IllegalArgumentException(
-          "Expected a function call, but got: %s".formatted(assistantMessage));
-    }
-
-    final WeatherMethod.Request arguments =
-        parseJson(functionCall.getArguments(), WeatherMethod.Request.class);
-    final WeatherMethod.Response weatherMethod = WeatherMethod.getCurrentWeather(arguments);
+    // 1. Define the function
+    final List<OpenAiTool<?>> tools =
+        List.of(
+            new OpenAiTool<>("weather", WeatherMethod.Request.class)
+                .setDescription("Get the weather for the given location")
+                .setFunction(WeatherMethod::getCurrentWeather));
 
-    messages.add(OpenAiMessage.tool(weatherMethod.toString(), functionCall.getId()));
+    // 2. Assistant calls the function
+    final var request = new OpenAiChatCompletionRequest(messages).withToolsExecutable(tools);
+    final OpenAiChatCompletionResponse response = client.chatCompletion(request);
 
-    // Send back the results, and the model will incorporate them into its final response.
-    return OpenAiClient.forModel(GPT_4O_MINI).chatCompletion(request.withMessages(messages));
-  }
+    // 3. Execute the tool call for given tools
+    final OpenAiAssistantMessage assistantMessage = response.getMessage();
+    final var toolResults = OpenAiTool.execute(tools, assistantMessage);
 
-  @Nonnull
-  private static <T> T parseJson(@Nonnull final String rawJson, @Nonnull final Class<T> clazz) {
-    try {
-      return JACKSON.readValue(rawJson, clazz);
-    } catch (JsonProcessingException e) {
-      throw new IllegalArgumentException("Failed to parse tool call arguments: " + rawJson, e);
-    }
-  }
+    // 4. Return the results so that the model can incorporate them into the final response.
+    messages.add(assistantMessage);
+    messages.addAll(toolResults.getMessages());
```

Comment:
Why a separate type `OpenAiTool.Execution` then `getMessages()` ?

Reply:
In which use-case? I don't see it

Reply:
Logging the tool execution and result would be a use-case.